### PR TITLE
feat: add the telegraf pod in container

### PR DESCRIPTION
**Pull request by kiragoo**

### CODE REVIEW

**Comment by Rory-Z** [comment]
We should add more test cases for telegraf

File: pkg/service/generator.go [hunk 1]
```diff
@@ -98,6 +109,36 @@ func generateStatefulSetDef(emqx v1beta1.Emqx) *appsv1.StatefulSet {
...
+	telegrafTemplate := emqx.GetTelegrafTemplate()
+
+	telegrafConfName := fmt.Sprintf("%s-%s", emqx.GetName(), "telegraf-config")
+	logName := fmt.Sprintf("%s-%s", emqx.GetName(), "log")
+
+	containerForTelegraf := corev1.Container{
+		Name:            telegrafTemplate.GetName(),
+		Image:           telegrafTemplate.GetImage(),
+		ImagePullPolicy: "IfNotPresent",
+		VolumeMounts: []corev1.VolumeMount{
```
[the remark this answers is not in the document]
Comment (kiragoo) [reply pkg/service/generator.go:119]: Already merged the function in `generateContainerForTelegraf()`.

File: pkg/service/generator.go [hunk 2]
```diff
@@ -586,6 +646,23 @@ func generateVolume(emqx v1beta1.Emqx, sts *appsv1.StatefulSet) *appsv1.Stateful
 	return sts
 }
 
+func generateVolumeForTelegraf(sts *appsv1.StatefulSet, name string) *appsv1.StatefulSet {
+	sts.Spec.Template.Spec.Volumes = append(
+		sts.Spec.Template.Spec.Volumes,
+		corev1.Volume{
```
[the remark this answers is not in the document]
Comment (kiragoo) [reply pkg/service/generator.go:652]: Already merged the function in `generateContainerForTelegraf()`.

File: apis/apps/v1beta1/telegraf_type.go [hunk 3]
```diff
@@ -0,0 +1,16 @@
+package v1beta1
+
+//+kubebuilder:object:generate=true
+type TelegrafTemplate struct {
+	Name  *string `json:"name,omitempty"`
```
Comment (Rory-Z) [inline apis/apps/v1beta1/telegraf_type.go:5]: Why use pointers instead of value references ?

File: apis/apps/v1beta1/telegraf_type.go [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+	Image *string `json:"image,omitempty"`
+	Conf  *string `json:"conf,omitempty"`
```
Comment (Rory-Z) [inline apis/apps/v1beta1/telegraf_type.go:7]: I suggest we add another `Resource` field

File: pkg/service/generator.go [hunk 5]
```diff
@@ -497,6 +538,25 @@ func generateConfigMapForModules(emqx v1beta1.Emqx, sts *appsv1.StatefulSet) (*c
 	return cm, sts
 }
 
+func generateConfigMapForTelegraf(emqx v1beta1.Emqx, sts *appsv1.StatefulSet) (*corev1.ConfigMap, *appsv1.StatefulSet) {
```
Comment (Rory-Z) [inline pkg/service/generator.go:541]: This function doesn't seem to have a useful `StatefulSet` argument, meybe we should merge `generateVolumeForTelegraf` in this

File: pkg/service/generator.go [hunk 6]
```diff
@@ -497,6 +538,25 @@ func generateConfigMapForModules(emqx v1beta1.Emqx, sts *appsv1.StatefulSet) (*c
...
+	telegrafConfName := fmt.Sprintf("%s-%s", emqx.GetName(), "telegraf-config")
+	telegrafConfData := emqx.GetTelegrafTemplate().Conf
```
Comment (Rory-Z) [inline pkg/service/generator.go:543]: Is there a problem here if `Conf` is not configured?

File: controllers_suite/suite_test.go [hunk 7]
```diff
@@ -272,6 +272,37 @@ func generateEmqxNamespace(namespace string) *corev1.Namespace {
 // Full
 func generateEmqxBroker(name, namespace string) *v1beta1.EmqxBroker {
 	storageClassName := "standard"
+	telegrafConf :=
+		`
```
Comment (Rory-Z) [inline controllers_suite/suite_test.go:276]: Is a space required ?

File: controllers_suite/suite_test.go [hunk 8]
```diff
@@ -272,6 +272,37 @@ func generateEmqxNamespace(namespace string) *corev1.Namespace {
...
+	[global_tags]
+	  instanceID = "emqx"
+
+	[agent]
+	  interval = "60s"
+      flush_jitter = "5s"
```
Comment (Rory-Z) [inline controllers_suite/suite_test.go:282]: Does it need to be aligned

**Pull request closed without merging**